art/attacks/inference/membership_inference.py

```diff
@@ -0,0 +1,346 @@
+        return [1 if p == y[index] else 0 for index, p in enumerate(predictions)]
+
+
+class MembershipInferenceBlackBox(InferenceAttack):
+    """
+        Implementation of a learned black-box membership inference attack.
+
+        This implementation can use as input to the learning process probabilities/logits or losses,
+        depending on the type of model and provided configuration.
+    """
+    _estimator_requirements = [BaseEstimator]
+
+    def __init__(self, classifier: Classifier, input_type: Optional[str] = 'prediction',
+                 attack_model_type: Optional[str] = 'nn', attack_model: Optional[Classifier] = None):
+        """
+        Create a MembershipInferenceBlackBox attack instance.
+
+        :param classifier: Target classifier.
+        :param attack_model_type: the type of default attack model to train, optional. Should be one of `nn` (for neural
+                                  network, default), `rf` (for random forest) or `gb` (gradient boosting). If
+                                  `attack_model` is supplied, this option will be ignored.
+        :param input_type: the type of input to train the attack on. Can be one of: 'prediction' or 'loss'. Default is
+                           `prediction`. Predictions can be either probabilities or logits, depending on the return type
+                           of the model.
+        :param attack_model: The attack model to train, optional. If none is provided, a default model will be created.
+        """
+        super(MembershipInferenceBlackBox, self).__init__(estimator=classifier)
+
+        if input_type not in ['prediction', 'loss']:
+            raise ValueError("Illegal value for parameter `input_type`.")
+
+        self.input_type = input_type
+
+        if attack_model:
+            if ClassifierMixin not in type(attack_model).__mro__:
+                raise TypeError("Attack model must be of type Classifier.")
+            self.attack_model = attack_model
+            self.default_model = False
+            self.attack_model_type = None
+        else:
+            self.default_model = True
+            self.attack_model_type = attack_model_type
+            if attack_model_type == 'nn':
+                if input_type == 'prediction':
+                    self.attack_model = MembershipInferenceAttackModel(classifier.nb_classes)
+                else:
+                    self.attack_model = MembershipInferenceAttackModel(classifier.nb_classes, 1)
+                self.epochs = 100
+                self.bs = 100
+                self.lr = 0.0001
+            elif attack_model_type == 'rf':
+                self.attack_model = RandomForestClassifier()
+            elif attack_model_type == 'gb':
+                self.attack_model = GradientBoostingClassifier()
+            else:
+                raise ValueError("Illegal value for parameter `attack_model_type`.")
```

Comment:
After setting all attributes we should run method `_check_params` which contains type and value checks for each attribute if possible. A method `_check_params` needs to be implemented by this class.
```suggestion
                raise ValueError("Illegal value for parameter `attack_model_type`.")
        self._check_params()
```